crates/dbsp/src/storage/dirlock/mod.rs

```diff
@@ -0,0 +1,160 @@
+//! Makes sure we don't accidentally run multiple instances of the program
+//! using the same data directory.
+
+use std::fs::{File, OpenOptions};
+use std::io::{self, Error as IoError, Read, Write};
+use std::os::fd::AsRawFd;
+use std::path::{Path, PathBuf};
+use sysinfo::{Pid, System};
+
+use crate::storage::backend::StorageError;
+
+#[cfg(test)]
+mod test;
+
+/// Check whether a process exists, used to determine whether a pid file is
+/// stale.
+fn process_exists(pid: u32) -> bool {
+    let s = System::new_all();
+    s.process(Pid::from(pid as usize)).is_some()
+}
+
+/// An instance of a PID file.
+///
+/// The PID file is removed from the FS when the instance is dropped.
+#[derive(Debug)]
+pub struct LockedDirectory {
+    base: PathBuf,
+    pid: Pid,
+}
+
+impl Drop for LockedDirectory {
+    fn drop(&mut self) {
+        let pid_file = self.base.join(LockedDirectory::LOCKFILE_NAME);
+        if pid_file.exists() {
+            log::trace!("Removing pidfile: {}", pid_file.display());
+            std::fs::remove_file(&pid_file).unwrap();
+        }
+    }
+}
+
+/// A guard that holds an exclusive lock on a file.
+///
+/// The lock is dropped once the guard is dropped.
+/// This avoids concurrency issues when two pipelines are started at the same
+/// time.
+struct FlockGuard(File);
+
+impl FlockGuard {
+    fn new(file: File) -> Result<FlockGuard, IoError> {
+        #[cfg(unix)]
+        {
+            let r = unsafe { libc::flock(file.as_raw_fd(), libc::LOCK_EX | libc::LOCK_NB) };
+            if r != 0 {
+                return Err(io::Error::last_os_error());
+            }
+        }
+        Ok(FlockGuard(file))
+    }
+}
+impl Drop for FlockGuard {
+    fn drop(&mut self) {
+        #[cfg(unix)]
+        {
+            unsafe {
+                libc::flock(self.0.as_raw_fd(), libc::LOCK_UN);
+            }
+        }
+    }
+}
+
+impl LockedDirectory {
+    const LOCKFILE_NAME: &'static str = "feldera.pidlock";
+
+    fn with_pid<P: AsRef<Path>>(base_path: P, pid: Pid) -> Result<LockedDirectory, StorageError> {
+        let pid_str = pid.to_string();
+        let pid_file = base_path.as_ref().join(LockedDirectory::LOCKFILE_NAME);
+
+        let mut guard = if pid_file.exists() {
+            // we set create(true) for both branches, to avoid concurrency issues when two
+            // pipelines are started at the same time.
+            let file = OpenOptions::new()
+                .read(true)
+                .write(true)
+                .create(true)
+                .open(&pid_file)?;
+            let mut guard = FlockGuard::new(file)?;
+
+            // From here on we have a lock on the file:
+            let mut contents = String::new();
+            guard.0.read_to_string(&mut contents)?;
+
+            let old_pid = contents.trim().parse::<u32>();
+            if let Ok(old_pid) = old_pid {
+                if process_exists(old_pid) {
+                    return Err(StorageError::StorageLocked(old_pid, pid_file));
```

Comment:
We'll only get this error in the weird case where some other PID has "taken" the lock but released the FlockGuard (since we got the FlockGuard if we made it here).